server/internal/router/route.go

```diff
@@ -0,0 +1,132 @@
+package router
+
+import (
+	"fmt"
+	"log"
+	"sync"
+)
+
+type routeMap struct {
+	// m maps a model to a list of engine IPs.
+	m map[model][]string
+	// engines is the collections of engine IPs.
+	engines map[string]struct{}
+
+	// mu protects m and engines.
+	mu sync.Mutex
+}
+
+func newRouteMap() *routeMap {
+	return &routeMap{
+		m:       make(map[model][]string),
+		engines: make(map[string]struct{}),
+	}
+}
+
+func (r *routeMap) getRoute(model model) []string {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	return r.m[model]
```

Comment:
I'm not sure about the concurrency guarantee there that a caller will access the slice outside of the mutex lock, and `addRoute` might be called.